File: test/keyboard_hint_test.rb
require_relative 'test_helper'

# The on-screen keyboard is a beginner's crutch: never there from the start,
# brought up when a note is clearly not being found — several wrong keys for
# it, or a long wait over it — with that note lit.
#
# two-measures.xml is one whole note per bar, C4 then D4, right hand.
class KeyboardHintTest < CapybaraTestBase
  def setup
    page.driver.set_cookie('test-env', 'true')
    # The strip fades in, and under a parked clock that fade can sit at its
    # first frame — opacity 0, which Capybara takes for not visible. Reduced
    # motion drops it: the strip is there the moment x-show reveals it.
    emulate_media([{ name: 'prefers-reduced-motion', value: 'reduce' }])
  end

  def teardown
    emulate_media([])
    super
  end

  def test_wrong_keys_for_a_note_bring_the_keyboard_up_with_that_note_lit
    open_two_measures
    # Clock held, so that the three keys alone bring it up, not the wait.
    with_clock_control do
      play_note('E4')
      play_note('F4')
      advance_clock(100)
      assert_keyboard_down

      play_note('G4')
      # Alpine reveals from a timer of its own.
      advance_clock(100)
      assert_selector '.pt-keyhint__chip', text: 'do4'
      assert_selector '.pt-keyhint__hand', text: 'MD'
      assert_selector '.pt-keyhint__key.is-owed[data-midi="60"]', text: 'do'
      assert_selector '.pt-keyhint__key.is-owed', count: 1

      # Found: the keyboard moves on to the next note.
      play_note('C4')
      assert_selector '.pt-keyhint__key.is-owed[data-midi="62"]', text: 'ré'
      assert_selector '.pt-keyhint__chip', text: 'ré4'
    end
  end

  # A lit white key must not climb over the black keys that sit across it.
  def test_the_black_keys_stay_on_top_of_a_lit_white_key
    open_two_measures
    3.times { play_note('G4') }
    assert_selector '.pt-keyhint__key.is-owed[data-midi="60"]'

    on_top = page.evaluate_script(<<~JS)
      (() => {
        const black = document.querySelector('.pt-keyhint__key[data-midi="61"]').getBoundingClientRect()
        return document.elementFromPoint(black.left + 2, black.top + black.height / 2)?.closest('.pt-keyhint__key')?.dataset.midi
      })()
    JS
    assert_equal '61', on_top
  end

  # Training replays the measure a beat after its last note: the keyboard
  # follows the cursor back to the note owed again, with no key pressed.
  def test_it_follows_the_cursor_through_a_training_repetition
    open_two_measures
    click_on 'Mode Entraînement'
    assert_text 'Mode Entraînement Actif'
    3.times { play_note('G4') }
    assert_selector '.pt-keyhint__key.is-owed[data-midi="60"]'

    play_note('C4')
    assert_selector '.pt-keyhint__key.is-owed[data-midi="60"]'
    assert_selector 'svg circle.repeat-indicator.filled', count: 0
    play_note('C4')
    assert_selector 'svg circle.repeat-indicator.filled', count: 1
    assert_selector '.pt-keyhint__key.is-owed[data-midi="60"]'
  end

  def test_a_long_wait_over_a_note_brings_the_keyboard_up
    open_two_measures
    with_clock_control do
      play_note('C4')
      advance_clock(5000)
      assert_keyboard_down

      advance_clock(4000)
      assert_selector '.pt-keyhint__key.is-owed[data-midi="62"]'
    end
  end

  def test_the_cross_puts_it_away_for_the_visit
    open_two_measures
    3.times { play_note('G4') }
    assert_selector '.pt-keyhint'

    find('.pt-keyhint__close').click
    assert_no_selector '.pt-keyhint'

    3.times { play_note('G4') }
    play_note('C4')
    3.times { play_note('G4') }
    assert_no_selector '.pt-keyhint'
  end

  # Feedback b7682019: the last note played, the cursor goes back to the top
  # under the results, and a wait over that first note is not hesitation. The
  # keyboard goes down with the piece and stays down, through the results and
  # after them, until the next run gives it a reason.
  def test_it_goes_down_with_a_finished_piece_and_stays_down
    open_two_measures
    3.times { play_note('G4') }
    assert_selector '.pt-keyhint'

    play_note('C4')
    play_note('D4')
    assert_text 'Partition terminée'
    assert_back_at_the_top

    with_clock_control do
      assert_stays_down_through_the_results

      3.times { play_note('G4') }
      advance_clock(100)
      assert_selector '.pt-keyhint__key.is-owed[data-midi="60"]'
    end
  end

  # A picked passage done opens its results from inside its last note, before
  # the keyboard has had that key: it must not count it as the next go begun.
  def test_it_does_not_come_up_over_a_finished_passage
    open_two_measures
    click_on 'Mode Entraînement'
    assert_text 'Mode Entraînement Actif'
    click_on '🔁 Boucle'
    click_measure(1)
    assert_text 'cliquez sur la dernière mesure du passage'
    click_measure(2)
    assert_text 'Mesures 1 à 2'
    3.times do
      play_note('C4')
      assert_selector 'svg rect.measure-click-area.selected[data-measure-index="1"]'
      play_note('D4')
      assert_back_at_the_top
    end
    assert_text 'Entraînement terminé'

    with_clock_control { assert_stays_down_through_the_results }
  end

  # Started from a bar further on, a run reaches the end with no results to
  # show, but it is over all the same.
  def test_it_does_not_come_up_after_a_run_ended_without_results
    open_two_measures
    click_measure(2)
    play_note('D4')
    assert_no_text 'Partition terminée'
    assert_back_at_the_top

    with_clock_control do
      advance_clock(20_000)
      assert_keyboard_down
    end
  end

  # Strict mode asks for notes on the metronome's time, not the player's: the
  # keyboard has no place there, and keys tried while it is selected do not
  # count toward the next mode either.
  def test_strict_mode_never_brings_it_up
    open_two_measures
    click_on '⏱ Mode strict'
    3.times { play_note('G4') }
    assert_no_selector '.pt-keyhint'

    click_on 'Libre'
    play_note('G4')
    assert_no_selector '.pt-keyhint'
  end

  private

  def open_two_measures
    visit '/score.html?url=/test-fixtures/two-measures.xml'
    wait_for_score_render(2)
  end

  # The beat after the last note, which takes the cursor back to the top.
  # Waited for before the clock is parked: parked ahead of it, a 20 s advance
  # was seen not to reach the wait the keyboard counts from there.
  def assert_back_at_the_top
    assert_no_selector 'svg g.vf-notehead.played-note'
  end

  def emulate_media(features)
    page.driver.browser.page.command('Emulation.setEmulatedMedia', features: features)
  end

  # Down the way x-show puts it, rather than merely not visible: a strip
  # caught mid-fade would pass assert_no_selector.
  def assert_keyboard_down
    assert_selector '.pt-keyhint[style*="display: none"]', visible: :all
  end

  # However long the results are read, and after they are closed. Inside
  # with_clock_control.
  def assert_stays_down_through_the_results
    advance_clock(20_000)
    assert_keyboard_down
    within('dialog.pt-result-dialog') { trigger_click_on 'Fermer' }
    advance_clock(20_000)
    assert_keyboard_down
  end
end

File: test/js/keyboardHint.test.js
import { describe, it, expect } from 'vitest'
import {
  hintStep,
  initialHint,
  keyboardRange,
  keyLayout,
  HINT_WRONG_NOTES,
  HINT_HESITATION_MS,
  HINT_FLUENT_NOTES,
  HINT_FLUENT_MS,
} from '../../public/js/keyboardHint.js'

const run = (events, state = initialHint()) => events.reduce(hintStep, state)
const tick = (key, now, eligible = true) => ({ type: 'tick', key, now, eligible })
const press = (now) => ({ type: 'press', now })
const wrong = (now) => ({ type: 'wrong', now })

// Reaches the keyboard being up, through wrong notes on the first note.
const revealed = () =>
  run([tick('0:0', 0), ...Array.from({ length: HINT_WRONG_NOTES }, (_, i) => wrong(100 + i))])

describe('hintStep', () => {
  it('stays down while the player finds the notes', () => {
    const state = run([tick('0:0', 0), press(500), tick('0:0.25', 500), press(1500), tick('0:0.5', 1500)])
    expect(state.visible).toBe(false)
  })

  it('comes up after several wrong keys for the same note', () => {
    const almost = run([tick('0:0', 0), ...Array.from({ length: HINT_WRONG_NOTES - 1 }, (_, i) => wrong(i))])
    expect(almost.visible).toBe(false)
    expect(hintStep(almost, wrong(10)).visible).toBe(true)
  })

  it('counts wrong keys per note, not per piece', () => {
    const state = run([
      tick('0:0', 0),
      wrong(10),
      wrong(20),
      press(30),
      tick('0:0.25', 30),
      wrong(40),
      wrong(50),
    ])
    expect(state.visible).toBe(false)
  })

  it('comes up when a note is waited on too long', () => {
    const waiting = run([tick('0:0', 0), press(0), tick('0:0.25', 0)])
    expect(hintStep(waiting, tick('0:0.25', HINT_HESITATION_MS - 1)).visible).toBe(false)
    expect(hintStep(waiting, tick('0:0.25', HINT_HESITATION_MS)).visible).toBe(true)
  })

  it('does not take a score left open before the first key for hesitation', () => {
    expect(run([tick('0:0', 0), tick('0:0', HINT_HESITATION_MS * 5)]).visible).toBe(false)
    // …and the wait starts with the first key, not with the page.
    const state = run([tick('0:0', 0), wrong(HINT_HESITATION_MS * 5), tick('0:0', HINT_HESITATION_MS * 5 + 10)])
    expect(state.visible).toBe(false)
  })

  it('does not count a wait while nothing is asked of the player', () => {
    const state = run([
      tick('0:0', 0),
      press(0),
      tick('0:0', HINT_HESITATION_MS * 2, false),
      tick('0:0', HINT_HESITATION_MS * 2 + 10),
    ])
    expect(state.visible).toBe(false)
  })

  it('does not count a wait after a mode change until the next key', () => {
    const state = run([tick('0:0', 0), press(0), { type: 'rest' }, tick('0:0', HINT_HESITATION_MS * 2)])
    expect(state.visible).toBe(false)
  })

  it('starts over at the end of a run, as on a piece just opened, the ✕ kept', () => {
    expect(hintStep(revealed(), { type: 'restart' })).toEqual(initialHint())
    expect(run([{ type: 'dismiss' }, { type: 'restart' }], revealed()).dismissed).toBe(true)
  })

  it('goes away once the notes come promptly and clean for a stretch', () => {
    // The note it came up for, found at last, is not part of the stretch.
    let state = hintStep(revealed(), tick('0:1', 1000))
    for (let i = 1; i < HINT_FLUENT_NOTES; i++) state = hintStep(state, tick(`1:${i}`, 1000 + 1000 * i))
    expect(state.visible).toBe(true)
    state = hintStep(state, tick('2:0', 1000 + 1000 * HINT_FLUENT_NOTES))
    expect(state.visible).toBe(false)
  })

  it('starts the stretch over on a slow or fumbled note', () => {
    let state = revealed()
    let now = 0
    const step = (key, gap) => {
      now += gap
      state = hintStep(state, tick(key, now))
    }
    step('0:1', 1000)
    for (let i = 1; i < HINT_FLUENT_NOTES; i++) step(`1:${i}`, 1000)
    step('2:0', HINT_FLUENT_MS + 1)
    expect(state.visible).toBe(true)
    expect(state.streak).toBe(0)
    step('2:1', 100)
    expect(state.streak).toBe(1)
    state = hintStep(state, wrong(now))
    step('2:2', 100)
    expect(state.streak).toBe(0)
  })

  it('stays away for the visit once dismissed', () => {
    let state = hintStep(revealed(), { type: 'dismiss' })
    expect(state.visible).toBe(false)
    state = run([tick('2:0', 0), wrong(1), wrong(2), wrong(3), tick('2:0', HINT_HESITATION_MS * 3)], state)
    expect(state.visible).toBe(false)
  })
})

describe('keyboardRange', () => {
  it('spans whole octaves around the notes', () => {
    expect(keyboardRange([50, 62, 79])).toEqual({ low: 48, high: 83 })
  })

  it('draws at least two octaves, growing toward middle C', () => {
    expect(keyboardRange([62, 67])).toEqual({ low: 48, high: 71 })
    expect(keyboardRange([74, 79])).toEqual({ low: 60, high: 83 })
    expect(keyboardRange([40, 45])).toEqual({ low: 36, high: 59 })
  })

  it('stops at the ends of a piano', () => {
    expect(keyboardRange([21, 108])).toEqual({ low: 21, high: 108 })
  })
})

describe('keyLayout', () => {
  it('places each black key over the gap after the white keys before it', () => {
    const { keys, whites } = keyLayout({ low: 60, high: 71 })
    expect(whites).toBe(7)
    expect(keys.filter((k) => k.black).map((k) => [k.midi, k.at])).toEqual([
      [61, 1],
      [63, 2],
      [66, 4],
      [68, 5],
      [70, 6],
    ])
  })
})
